HW2/P4/P4.txt

```diff
@@ -0,0 +1,30 @@
+No Threading Module (Single Threaded) 
+  3.34195303917 seconds for 10 filter passes.
+
+Sample run is representative of multiple runs
+
+Threading Module
+
+  1 Thread
+  3.34847378731 seconds for 10 filter passes.
+
+  2 Threads
+  1.75988101959 seconds for 10 filter passes.
+
+  4 Threads
+  0.937849998474 seconds for 10 filter passes.
+
+  Sample runs are representative of multiple runs
+
+
+Implementation
+
+  To make sure that threads do not collide with each other, I had a pair of semaphores (lower, upper) for each thread, all initialized with a count of 1.
+
+  When thread (n) is about to start an iteration, it acquires the upper semaphore of the (n-1) thread and the lower semaphore of the (n+1) thread.  Once it finishes the iteration, it increments its own lower and upper semaphores.
```

Comment:
why does it need to increment the lower *and* upper semaphores? shouldn't a single thread finishing just have to increment it's own semaphore, indicating thread i has finished iteration n?